### ISSUE TITLE
request: provide (Windows) binary releases

### ISSUE THREAD

**Issue by AntonOks** [axboe/fio#1128]
First, thanks for this nice tool. I like it a lot and use it every couple of months to test different disks and storage systems.

because the (Windows) binaries are only available either as artifacts in **ci.appveyor.com** or in "google drive" and their download links are changing with every release, it is currently almost impossible to automate FIO **updates** in various Windows package management solutions like [chocolatey ](https://chocolatey.org/) or [scoop](https://scoop.sh/) .

due to this, Windows admins have to install and, even worse, keep FIO up2date manually, which obviously results in either high efforts or in an unpleasant FIO experience and therefore a low acceptance within automation, frameworks / workflows...

_Examples for broken or outdated FIO links in the most popular Windows management tools_:
- https://chocolatey.org/packages/fio - still on v3.13
- https://github.com/ScoopInstaller/Main/issues/1598 - broken links

It would be great if FIO would provide (Windows) **binary releases on GitHub** as so many other tools do.
With this, it would be a breeze for (Windows) admins and/or package managers to automatically roll out and use FIO within their performance test frameworks and workflows. FIOs use and acceptance would for sure rise (in the Windows world) :)
 

Unfortunately I have ZERO appveyor knowledge...  even if, based on the available documentation, if seems not to be too hard to implement a artifact push to GitHub:
- [Publishing artifacts to GitHub Releases](https://www.appveyor.com/docs/deployment/github/)

@sitsofe as you did some great additions to [appveyor.yml](https://github.com/axboe/fio/commits/master/.appveyor.yml) in the past, I hope you would so kind ans could again help with a "deploy to GitHub" pull?

**Comment by sitsofe** [comment]
@AntonOks I have noticed we're getting more requests to pre-package binaries recently which I think reflects fio's changing user base. From what I've read, @axboe is fine with the project shipping binaries so long as someone steps up to do the work...

I'm not actually a professional Windows developer - I currently don't have a Windows install which makes things even more interesting... What I'd prefer is for someone with more experience and an interest in furthering fio on that platform to come on board and help us. In my experience the easier and more official things become the more demands are made by end users so it would be good to have such a person already in place. On a personal note I really would like to see Windows issues like #1073 etc. solved before we open the floodgates...

I'll label this up an enhancement request and that's we're happy to review patches on this.

**Comment by axboe** [comment]
Just to confirm what @sitsofe said - I'm totally fine with having windows binaries generated in a predictable fashion so that folks can use them. I don't have access to Windows either and it's really outside of my area of expertise, so it's not something that I'm personally going to be working on. But if someone steps up and does the work, I'd be happy to help get it integrated into the build system.

**Comment by AntonOks** [comment]
@sitsofe 

> From what I've read, @axboe is fine with the project shipping binaries so long as someone steps up to do the work...
good to know

> What I'd prefer is for someone with more experience and an interest in furthering fio on that platform to come on board and help us.
> In my experience the easier and more official things become the more demands are made by end users so it would be good to have such a person already in place

Well, currently FIO makes it hard to get more love from Windows admins and/or Dev's, see my initial comments. So it's the chicken-egg problem here. We want Windows dev's to help and maintain things but on the other side FIO makes it, by intention as you say below, hard to get the **_already build_** Appveyor artifacts (aka the Windows MSI packages!). 

> ... before we open the floodgates...

So the current situation "is by intention", right?

> I'll label this up an enhancement request and that we're happy to review patches on this.

**Thanks**. As I said, I unfortunately I have ZERO Appveyor know-how. Otherwise I would have send a pull already  :|

**Comment by AntonOks** [comment]
> Just to confirm what @sitsofe said - I'm totally fine with having windows binaries generated in a predictable fashion so that folks can use them. I don't have access to Windows either and it's really outside of my area of expertise, so it's not something that I'm personally going to be working on. But if someone steps up and does the work, I'd be happy to help get it integrated into the build system.

@axboe Thanks. Great to know!
If I'm not wrong, the Windows binaries are build today already, right? That's why they can be grabbed from the Appveyor artifacts, as stated in the README.

High level, my "request" is to just extend the existing **_appveyor.yml_** so that the already build artifacts (Windows MSI binaries) are uploaded to the GitHub releases area. So no personal Windows system or such know-how is needed, but "only" some Appveyor foo... but maybe I misunderstand the documentation...

**Comment by axboe** [comment]
> If I'm not wrong, the Windows binaries are build today already, right? That's why they can be grabbed from the Appveyor artifacts, as stated in the README.

Right, it's really just a matter of making the builds easier to find/reach as far as I can tell. They are already generated as part of the CI testing.

**Comment by sitsofe** [comment]
> So the current situation "is by intention", right?

No. The current situation is because we have nascent Windows support and growing Windows CI but we don't yet have Windows developers regularly contributing/answering (Windows) questions. I worked on the Windows CI in my spare time to help us stop breaking Windows builds by accident even though I don't actually use Windows myself!

@AntonOks What I did a poor job of saying is: I would love for someone else to fulfill your request because I don't have time to work on it but I'm happy to review work that solves it. I totally get that the person submitting the work does not have to be you but since you named me explicitly I want to make it clear that there may be others with stronger motivation to solve it quicker than I will. 

@axboe Releases seem to be special and are maybe linked to tags. I would assume we want to only upload those builds to GitHub that match a tagged fio (otherwise every CI build would have to have a "Release")? I think the work would revolve around changing things so the CI knows when it is building a tagged release off the main repo (to stop things like pull requests doing something malicious) and then actually uploading the MSI to the matching GitHub release with the right keys etc.
